부산대 FE 안형찬 - 1단계 선물하기 메인 API 구현하기

hyeongchanan opened the pull request:
멘토님 이번에도 잘 부탁드립니다! 

-파일 구조 리팩토링
-목 서버를 이용해 상품 테마 정보 표시
-상품 랭킹 정보 표시
-선택에 따라 랭킹 정보 다르게 표시
-로딩화면과 에러 화면 구현

hyeongchanan committed:
0단계 및 피드백 반영

추가 리펙토링

대소문자 통일 등 형식 통일

피드백 반영 완료

메모제이션, 변수 위치변경, 기타 오류, 사용하지 않는 코드 제거 등등

스타일 변경

지난번 2단계에서 잘못된 버전을 가져와서 3단계를 시작해버려서
다시 가져온후 머지

초기 세팅

초기세팅 및 README 업데이트

hyeongchanan committed:
api 사용으로 테마 표시

hyeongchanan committed:
상품 랭킹 정보 표시

api로 상품랭킹 정보 표시,
quary를 이용해 선택한 타입별로 표시
파일 위치 수정/ 스타일 수정 및 코드 분리

june-by commented on src/component/main/GiftRankingList.tsx:
```diff
@@ -0,0 +1,115 @@
+
+
+const GiftRankingList = ({ targetType, rankType }: GiftRankingListProps) => {
+    const [Ranking, setThemes] = useState<ProductItem[]>([]);
+    const [Loding, setLoding] = useState(true);
+    const [Error, setError] = useState(false);
+
+    useEffect(() => {
+        const fetchTheme = async () => {
+            const rankingUrl = `${baseRankingUrl}?targetType=${targetType}&rankType=${rankType}`
+            const Ranking = await getFromUrl(rankingUrl);
+            if (Ranking) {
+                setThemes(Ranking.data);
+
+            } else {
+                setError(true);
+            }
+            setLoding(false);
+        };
+
+        fetchTheme();
+    }, [targetType, rankType])
```
커스텀훅을 배웠으니까, 분리해보는건 어떨까요?

june-by commented on src/utils/getFromUrl.ts:
```diff
@@ -0,0 +1,12 @@
+  try {
+    const response = await axios.get(url);
+    return response.data;
+  } catch (error) {
+    console.error('Error:', error);
+    return null;
+  }
```
에러가 발생했을 때, null을 리턴하게되는데요.

이러면 사용처에서 null을 리턴받았을 때, 이게 에러때문에 null이 리턴된건지 아니면 데이터가 없어서 null이 리턴된건지 구분하기가 힘들어질 것 같아요

null을 리턴하기보다는 throw Error등을 통해 사용처에 에러가 발생했음을 알리고 사용처에서 각각 상황에 맞게 에러처리를 해주는게 좋아보여요

june-by commented on src/component/order/OrderCheck.tsx:
```diff
@@ -0,0 +1,88 @@
+
+  const [productCount, setProductCount] = useState(0);
+
+  useEffect(() => {
+    const total  = receivers.reduce((acc,receiver) => acc + receiver.count, 0);
+    setProductCount(total);
+  },[receivers])
```
receivers가 변화할 때, 실행되어야 하는 로직인데요!

productCount가 굳이 상태로 설정되고, 이를 useEffect로 설정해주어야할 이유가 있을까요?

제가 보았을 땐, 상태로선언하지 않고 useEffect를 사용하지 않아도 될 것 같아요

useEffect를 왜 사용하는지, 상태를 통해 생성되는 파생상태를 어떻게 관리할지 고민해보시면 좋을 것 같습니다 ㅎㅎ

> https://velog.io/@jay/you-might-need-useEffect-diet

요것도 한번 읽어봐주세요!

june-by commented on src/component/order/OrderCheck.tsx:
```diff
@@ -0,0 +1,88 @@
+
+
+  const handleOrder = () => {
+    senderNameInput.validate();
+
+    const isValid =
+      !senderNameInput.error
```
`senderNameInput.validate()`에서도 에러를 리턴하는것으로 보이는데, 굳이 함수호출과 값을 가져와서 검증하는부분을 나눠야 할까요??

june-by commented on src/component/main/GiftTheme.tsx:
```diff
@@ -0,0 +1,64 @@
+  useEffect(() => {
+    const fetchTheme = async () => {
+      const theme = await getFromUrl(themeUrl);
+      if (theme) {
+        setThemes(theme.data);
+
+      } else {
+        setError(true);
+      }
+      setLoding(false);
```
`getFromUrl` 함수 코멘트에도 남겨놓았지만, 에러가 발생했을 때, 로딩이 끝났을때가 직관적으로 느껴지지 않습니다 ㅠㅠ
null을 리턴하는 방식을 수정해주세요

june-by commented on src/component/main/GiftRankingList.tsx:
```diff
@@ -0,0 +1,115 @@
+        fetchTheme();
+    }, [targetType, rankType])
+
+    const { user } = useAuth();
+    const navigate = useNavigate();
+
+
+    const handleClickProduct = (item: ProductItem) => {
+        if (!user) {
+            navigate(`/login?redirect=/order?id=${item.id}`);
+        } else {
+            navigate(`/order?id=${item.id}`, { state: { item } });
+        }
+    };
+
+    const [isExpanded, setIsExpanded] = useState(false);
+
+    const visibleCount = isExpanded ? Ranking.length : GIFTLENGTH;
+
+    const shownProducts = Ranking.slice(0, visibleCount);
```
컴포넌트 내부에서 선언하는 상태, 변수, 훅 그리고 실행하는 `useEffect`가 일정한 순서대로 배치되면 훨씬 보기좋을 것 같아요
ex) useState -> 변수선언 -> 함수선언 -> useEffect

june-by reviewed (changes requested):
형찬님 안녕하세요~

먼저 빌드를 돌리면 에러가 발생해서 수정 부탁드립니다!

그리고 1단계 PR에 0단계 PR에 포함되어있던 커밋들이 들어있는데요.

그래서 이번 PR에 0단계 PR에서 반영한 내용이 포함되어있어 리뷰가 힘드네요 ㅠㅠ

단계가 끝나고 멘토가 스쿼시 머지를 통해 next-step 원본 레포에 반영을 하면, 그걸 기준으로 브랜치를 새로 만들어서 작업 부탁드립니다!

hyeongchanan committed:
Merge branch 'hyeongchanan' into upstram/hyeongchan1

hyeongchanan committed:
피드백 수정

1. useFetchFromUrl 을 생성해서 URL 에서 데이터를 받아오는 코드 분리
2. 에러 처리
3. 코드 위치 수정
4.불필요한 코드 삭제

hyeongchanan commented:
말씀해주신 부분들 수정했습니다!
useEffect 나 에러 처리같은 부분들도 더욱 신경쓰도록 하겠습니다. 빠르고 꼼꼼한 리뷰 감사드려요

그리고 커밋이 꼬이고 복잡하게해서.. 정말 죄송합니다! 다음부터는 피드백이 올때까지 기다리고 작업하겠습니다

hyeongchanan reviewed (commented):
피드백 받은 부분 수정 완료했습니다!
-커스텀 훅으로 분리
-불필요한 useEffect 및 코드 삭제
-에러 처리
-코드 위치 수정

june-by reviewed (approved):
코드가 많이 깔끔해졌네요 ㅎㅎ
고생하셨습니다 👍 
다음 단계 진행해주세요~ 🚀
